This readme describes the files in this directory

### TRACS manifests

- `sequencing/json_summary.tsv` output from IRODS call - not very useful  

- `sequencing/tracs_samples_manifest.csv` All samples sequenced under TRACS - the
IRODs data linked to sample identifier with extra fields to say what type of
sample (sweep/pick) and source adn study (LCL, Whiston, TRACS, Companion
animals)  


- `sequencing/tracs_single_colony_wgs_manifest.csv` The above but restricted to
WGS single colony picks and put in a format to pass to IRODS to pull reads.  

These latter two files are generated using the script
`R/sequencing_analysis/collate_sequence_metadata.R`

### Outputs of qc tools run on the TRACS samples

These were run on the samples listed in `tracs_single_colony_wgs_manifest.csv`

- `sequencing/bracken_summary.tsv`   
- `sequencing/checkm_summary.tsv`  
- `sequencing/summary_quast_report.tsv`  


### Final, QC'd collections of samples 

- `sequencing/final_qc.csv` Collated sample metadata and QC data as generated by
the script `bioinformatics/single_colony/scripts/sample_qc.qmd`  

The following files are lists of the files that have passed qc for downstream
analysis, gernerated by the script  `R/sequencing_analysis/pull_wgs_ecoli_and_kleb.R`
  
- `sequencing/tracs_wgs_ecoli_assemblies.txt`  
- `sequencing/tracs_wgs_ecoli_lanes.txt`  
- `sequencing/tracs_wgs_kleb_assemblies.txt`  
- `sequencing/tracs_wgs_kleb_lanes.txt`  
